### ISSUE TITLE
Razor pages and webapi

### ISSUE THREAD

**Issue by eByte23** [Azure/azure-functions-host#2557]
Hi,

Amazon release functionality in Lambda to enable use of razor pages and webapi as functions.
repo here:  https://github.com/aws/aws-lambda-dotnet
article https://aws.amazon.com/blogs/developer/serverless-asp-net-core-2-0-applications/

Would this be something that would be considered for a feature request and/or pr?

Thanks

**Comment by dersia** [comment]
Hi Elijah,

I just went through the article. What they are doing here is, using pre-compiled razor pages.

I'd like to know what you'd like to use this for. Especially in an serverless (therefore stateless) environment. Just to understand the usecase here.

Cheers,
Sia

Sent from my Windows 10 phone

**Comment by eByte23** [comment]
Yes sure its using pre compiled razor but it is still running the full app I believe piped through the host class.

Basic use case is for a website that maybe private, or small business user that maybe have a semi complex site that doesnt have large work time and is use by few people but maybe it will be used more people in future.

Thus giving you a very cheap and reliable infrastructure.

**Comment by eByte23** [comment]
Also its not just razor pages its webapi as well

**Comment by SimonLuckenuik** [comment]
Closest I have been from hosting WebAPI is the with the following approach: https://gist.github.com/SimonLuckenuik/2f35167e1df1d9ee82c17e01893e8dbc
It does work, but it involves using the Unit TestServer from Microsoft.AspNetCore.TestHost which is obviously not meant to be used in production...

**Comment by SimonLuckenuik** [comment]
Using Azure Functions V2, I was able to have Razor Pages running properly inside a Functions host using the TestServer approach: http://www.luckenuik.net/hosting-your-aspnet-core-razor-pages-inside-azure-functions/

This is a proof of concept, but if we develop something similar to TestServer of production grade (and proper tooling to ease the job), we should be able to leverage the full power of MVC APIs\Pages inside Azure Functions.

I do understand that Azure Functions is meant to target multiple languages and that this is a .NET only approach, but I believe there is a need for such an approach. Obviously, something built-in Azure Functions would be preferable...

I know there are some people out there against running AspNet Core MVC inside Azure Functions, but the complete discussion around "it will take more memory, increase start-up time and adds overhead" is meaningless in my view, at least for some scenarios:

- Startup time: we are already facing Cold Starts today, that's not going away, adding few milliseconds to initialize the MVC stack is irrelevant. With Azure Functions Premium plan, these concerns will be less relevant, I guess.

- Execution Cost: No matter how much memory you are using, for some apps it is still cheaper than having an App Service running 24/7.

- Ramp-up Cost: It is way cheaper to take the existing knowledge of developers and make them code as usual with MVC Controllers instead of ramping-up with Azure Functions.

- Reuse Opportunity: A lot of work was done by different projects to increase the AspNet Core middleware with interesting features (Ex: Authn\Authz), it is a shame that we cannot reuse that work easily with Azure Functions.

- Complexity: If you want to run any business application on Azure Functions, you will face complex code that might take longer than 50 milliseconds. It's not true, in my view, that Azure Functions was designed for small projects only, with quick "in and out" only.

**Comment by jovansnevress** [comment]
Helllllp cyber bullying!🤧

Johan Nevress
323-406-0557

**Comment by martonx** [comment]
Hi is there any news about this feature request?
It would be great to host small Asp.Net Core Web API-s with routing in Azure Functions.
